components/SerialServer/src/serial.c

```diff
@@ -45,6 +49,7 @@ typedef struct getchar_buffer {
 compile_time_assert(getchar_buf_sized, sizeof(client_buffer_t) == sizeof(Buf));
 
 typedef struct getchar_client {
+    unsigned int idx;
```

Comment:
I think `idx` is not needed and make things more complicated than necessary. The field`client_id` could simply become `batch`. The helper `getchar_index_from_badge()` would become `getchar_client_from_badge()`

```
static getchar_client_t getchar_clients[MAX_BADGE];

getchar_client_t *getchar_client_from_badge(int badge)
{
    for (int i = 0; i < ARRAY_ELEMENTS(getchar_clients); i++) {
        if (getchar_clients[i].badge == badge) {
            return &getchar_clients[i];
        }
    }
    return null;
}
 
void pre_init(void)
{
    [...]

    int j = 0
    for (int i = 0; i < num_getchar_clients; i++) {
            unsigned int badge = getchar_enumerate_badge(i);
            if (badge > MAX_BADGE) {
                ZF_LOGE("Client badge %d exceeds range 0->%d", badge, MAX_BADGE);
                continue;
            }
            getchar_client_t *c = &getchar_clients[j++]
            c->badge = badge;
            c->buf = getchar_buf(badge);
            c->last_head = -1;
        
    [...]
```

Reply:
I added your getchar_client_from_badge file. That's a good way to remove the index.